swig/StringArray_API_extensions.i

```diff
@@ -0,0 +1,281 @@
+     * num_elements are filled.
+     *
+     * @param num_elements Number of strings to store in the array.
+     * @param string_size The size of each string in the array.
+     */
+    void _allocate_strings(int num_elements, int string_size)
+    {
+        for (int i = 0; i < num_elements; ++i)
+        {
+            // Leave space for \0 terminator:
+            _array[i] = new (std::nothrow) char[string_size + 1];
+
+            // Check memory allocation:
+            if (! _array[i]) {
+                _release_strings();
+                throw std::bad_alloc();
+            }
+        }
+    }
+
+    /**
+     * Deletes the allocated strings.
+     */
+    void _release_strings() noexcept
+    {
+        std::for_each(_array.begin(), _array.end(), [](char* c) { delete[] c; });
+    }
+
+    const size_t _string_size;
+    std::vector<char*> _array;
+};
+
+
+
+
+
+
+
+    /**
+     * Return the pointer to the array of strings.
+     * Wrapped in Java into String[] automatically.
+     */
+    char **StringArrayHandle_get_strings(StringArrayHandle handle)
+    {
+        return reinterpret_cast<StringArray *>(handle)->data();
+    }
+
+    /**
+     * For the end user to extract a specific string from the StringArray object.
+     */
+    char *StringArrayHandle_get_string(StringArrayHandle handle, int feature_idx)
+    {
+        return reinterpret_cast<StringArray *>(handle)->getitem(feature_idx);
+    }
+
+    /**
+     * Free the StringArray object.
+     */
+    void StringArrayHandle_free(StringArrayHandle handle)
+    {
+        delete reinterpret_cast<StringArray *>(handle);
+    }
+
+    /**
+     * Retrieve the number of strings in the StringArray.
+     */
+    int StringArrayHandle_get_num_elements(StringArrayHandle handle)
+    {
+        return reinterpret_cast<StringArray *>(handle)->get_num_elements();
+    }
+
+    /**
+     * @brief Wraps LGBM_BoosterGetEvalNames.
+     *
+     * In case of success a new StringArray is created and returned,
+     * which you're responsible for freeing,
+     * @see StringArrayHandle_free().
+     * In case of failure such resource is freed and nullptr is returned.
+     * Check for that case with null (lightgbmlib) or 0 (lightgbmlibJNI).
+     *
+     * @param handle Booster handle
+     * @param eval_counts number of evaluations
+     *
+     * @return StringArrayHandle with the eval names (or nullptr in case of error)
+     */
+    StringArrayHandle LGBM_BoosterGetEvalNamesSWIG(BoosterHandle handle,
+                                                   int eval_counts)
+    {
+        std::unique_ptr<StringArray> strings(nullptr);
+
+        try {
+            // TODO: @reviewer, 128 was the chosen size, any particular reason for this constraint?
+            strings.reset(new StringArray(eval_counts, 128));
+        } catch (std::bad_alloc &e) {
+            LGBM_SetLastError("Failure to allocate memory."); // TODO: @reviewer - why is this not setting the message?
```

Comment:
sorry, can you elaborate - do you mean why the previous code prior to this PR was not doing this or do you mean that this is currently not working?  If you mean the previous implementation that's a good point, we probably should have done this.

Reply:
Hi :)

I mean that it is not setting the message. If you add a `throw std::bad_alloc()` in the try block you'll see that the output message from `LGBM_GetLastError` is the default, that everything is ok.
Is this because we're in a separate compilation unit and the message array is static on the original code? Either way, we should find a way to make it work.